index.js

```diff
@@ -37,58 +40,69 @@ function paginate(query, options, callback) {
     offset = 0;
     skip = offset;
   }
-  if (limit) {
-    let docsQuery = this.find(query)
+
+  if (limit > 0) {
+    docsQuery = this.find(query)
       .select(select)
       .sort(sort)
       .skip(skip)
       .limit(limit)
       .lean(lean);
+
     if (populate) {
-      [].concat(populate).forEach((item) => {
-        docsQuery.populate(item);
-      });
+      [].concat(populate).forEach((item) => docsQuery.populate(item));
     }
-    promises = {
-      docs: docsQuery.exec(),
-      count: this.count(query).exec()
-    };
-    if (lean && leanWithId) {
-      promises.docs = promises.docs.then((docs) => {
-        docs.forEach((doc) => {
+  }
+
+  promises = [
+    docsQuery ? docsQuery.exec() : Promise.resolve({}),
+    this.count(query).exec()
+  ];
+
+  let promise = new Promise((resolve, reject) => {
+    Promise.all(promises).then((data) => {
+      let docs = (limit > 0) ? data[0] : [];
+      let count = data[1];
+      let result = {
+        docs: docs,
+        total: count,
+        limit: limit
+      };
+      if (lean && leanWithId) {
+        result.docs = result.docs.map((doc) => {
           doc.id = String(doc._id);
+          return doc;
         });
-        return docs;
-      });
-    }
-  }
-  promises = Object.keys(promises).map((x) => promises[x]);
-  return Promise.all(promises).then((data) => {
-    let result = {
-      docs: data.docs,
-      total: data.count,
-      limit: limit
-    };
-    if (offset !== undefined) {
-      result.offset = offset;
-    }
-    if (page !== undefined) {
-      result.page = page;
-      result.pages = Math.ceil(data.count / limit) || 1;
-    }
-    if (typeof callback === 'function') {
-      return callback(null, result);
-    }
-    let promise = new Promise();
-    promise.resolve(result);
-    return promise;
+      }
+      if (offset !== undefined) {
+        result.offset = offset;
+      }
+      if (page !== undefined) {
+        result.page = page;
+        result.pages = Math.ceil(result.total / limit) || 1;
+      }
+      if (typeof callback === 'function') {
+        return callback(null, result);
+      }
+      resolve(result);
+    }, (error) => {
+      if (typeof callback === 'function') {
+        return callback(error, null);
+      }
+      reject(error);
+    });
   });
+
+  return promise;
 }
 
 /**
  * @param {Schema} schema
+ * use native ES6 promises verus mpromise
  */
 
+mongoose.Promise = global.Promise;
```

Comment:
I would be wary about this line. Some people like to set `mongoose.Promise` to their own promise implementation (e.g. bluebird). Overwriting it here would cause them lots of headaches.